api/src/main/java/javax/enterprise/inject/spi/builder/BeanBuilder.java

```diff
@@ -0,0 +1,369 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2015, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package javax.enterprise.inject.spi.builder;
+
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.inject.Default;
+import javax.enterprise.inject.Instance;
+import javax.enterprise.inject.spi.AnnotatedType;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanAttributes;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.enterprise.inject.spi.PassivationCapable;
+import javax.enterprise.util.TypeLiteral;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * This API is an helper to build a new {@link Bean} instance.
+ * CDI container must provides an implementation of this interface accessible by calling
+ * {@link javax.enterprise.inject.spi.AfterBeanDiscovery#beanBuilder()}
+ * The same BeanBuilder can be used multiple times in one AfterBeanDiscovery observer but it shouldn't be used
+ * after the end of observer invocation.
+ *
+ *
+ * @author Martin Kouba
+ * @author Antoine Sabot-Durand
+ * @param <T> the class of the bean instance
+ * @since 2.0
+ */
+public interface BeanBuilder<T> {
```

Comment:
Maybe we could use generics and have something similar to TCK's `ArchiveBuilder`:
https://github.com/cdi-spec/cdi-tck/blob/master/impl/src/main/java/org/jboss/cdi/tck/shrinkwrap/ArchiveBuilder.java#L165
https://github.com/cdi-spec/cdi-tck/blob/master/impl/src/main/java/org/jboss/cdi/tck/shrinkwrap/WebArchiveBuilder.java
Although it might look like over-engineering. I'm not sure whether there is some better solution...